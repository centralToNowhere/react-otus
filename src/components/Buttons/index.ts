export { ButtonGameControl } from "./ButtonGameControl";
export { FormButton } from "./FormButton";
