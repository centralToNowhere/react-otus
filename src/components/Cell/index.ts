export { Cell } from "./Cell";
export type { ICell, ICellProps } from "./Cell";
