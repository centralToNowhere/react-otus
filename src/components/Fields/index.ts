export { InputField } from "./InputField";
export { LabelField } from "./LabelField";
export { InputPatterns } from "./Properties";
export {
  fieldControlSlice,
  setCellSize,
  setMaxFieldWidth,
  setMaxFieldHeight,
  setCapacity,
  setSpeed,
  resetFieldControls
} from "./FieldControlRdx";
export type { IFieldControlState } from "./FieldControlRdx";
export { useOnChangeHandler, onBlurHandler } from "./FieldHandlers";
