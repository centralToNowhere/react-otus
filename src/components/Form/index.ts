export { FormContainer } from "./FormContainer";
export { FormElement } from "./FormElement";
export { FormField } from "./FormField";
export { FormGroup, FormGroupButtons } from "./FormGroup";
export { Form } from "./Form";
export type {
  FormContainerProps,
  IFieldProps,
  IButtonProps,
} from "./FormContainer";
export type { IFormProps } from "./Form";
