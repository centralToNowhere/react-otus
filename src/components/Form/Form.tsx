import React from "react";
import styled from "@emotion/styled";
import { COLORS } from "@/styles/ui-styled";
import { minCellSize } from "@/components/Cell";

type ControlCallback<T = void> = (value: T) => void;
type Callback<E, T> = (e: E) => T;
export type FormProps = typeof Form.defaultProps & {
  cellSize?: number;
  onCellSizeChange?: ControlCallback<string>;
  maxFieldWidth?: number;
  onMaxFieldWidthChange?: ControlCallback<string>;
  maxFieldHeight?: number;
  onMaxFieldHeightChange?: ControlCallback<string>;
  readonly cellsInRow?: number;
  readonly cellsInCol?: number;
  capacity?: number;
  onCapacityChange?: ControlCallback<string>;
  speed?: number;
  onSpeedChange?: ControlCallback<string>;
  onStart?: ControlCallback;
  onStop?: ControlCallback;
  onReset?: ControlCallback;
};

type InputProperties =
  | "fieldWidth"
  | "fieldHeight"
  | "cellSize"
  | "capacity"
  | "speed";

type IInputErrors = {
  [key in InputProperties]: {
    show: boolean;
    msg: string;
  };
};

interface IFormState {
  cellSizeString: string;
  capacityString: string;
  maxFieldWidthString: string;
  maxFieldHeightString: string;
  speedString: string;
  errors: IInputErrors;
}

export const isValidNumericString = (
  inputString: unknown
): inputString is number => {
  return (
    typeof inputString === "string" &&
    inputString !== "" &&
    !isNaN(Number(inputString))
  );
};

const isValidCellSizeString = (cellSizeString: unknown): boolean => {
  return (
    isValidNumericString(cellSizeString) &&
    Number(cellSizeString) >= minCellSize
  );
};

export const isValidPositiveNumericString = (str: unknown): boolean => {
  return isValidNumericString(str) && Number(str) > 0;
};

export const isValidNonNegativeNumericString = (str: unknown): boolean => {
  return isValidNumericString(str) && Number(str) >= 0;
};

const floatPattern = "([0-9]*[.])?[0-9]+";
const numberPattern = "[0-9]*";

export class Form extends React.Component<FormProps, IFormState> {
  static defaultProps = {
    onCellSizeChange: (value: string) => {
      console.log(value);
    },
    onMaxFieldWidthChange: (value: string) => {
      console.log(value);
    },
    onMaxFieldHeightChange: (value: string) => {
      console.log(value);
    },
    onCapacityChange: (value: string) => {
      console.log(value);
    },
    onSpeedChange: (value: string) => {
      console.log(value);
    },
    onStart: () => {
      // do nothing
    },
    onStop: () => {
      // do nothing
    },
    onReset: () => {
      // do nothing
    },
  };

  private timer: ReturnType<typeof setTimeout> | null;
  private formRef: React.RefObject<HTMLFormElement>;

  static inputDelay = 1000;

  constructor(props: FormProps) {
    super(props);

    this.state = {
      cellSizeString: String(this.props.cellSize),
      capacityString: String(this.props.capacity),
      maxFieldWidthString: String(this.props.maxFieldWidth),
      maxFieldHeightString: String(this.props.maxFieldHeight),
      speedString: String(this.props.speed),
      errors: {
        fieldWidth: {
          show: false,
          msg: "Expected non-negative number",
        },
        fieldHeight: {
          show: false,
          msg: "Expected non-negative number",
        },
        capacity: {
          show: false,
          msg: "Expected non-negative number",
        },
        cellSize: {
          show: false,
          msg: "Expected number >= 10",
        },
        speed: {
          show: false,
          msg: "Expected positive number",
        },
      },
    };

    this.formRef = React.createRef();
    this.timer = null;
  }

  validateFieldWidth = (fieldWidthString: unknown): boolean => {
    if (isValidNonNegativeNumericString(fieldWidthString)) {
      this.setErrorState("fieldWidth", false);
      return true;
    }

    this.setErrorState("fieldWidth", true);
    return false;
  };

  validateFieldHeight = (fieldHeightString: unknown): boolean => {
    if (isValidNonNegativeNumericString(fieldHeightString)) {
      this.setErrorState("fieldHeight", false);
      return true;
    }

    this.setErrorState("fieldHeight", true);
    return false;
  };

  validateCapacity = (capacityString: unknown): boolean => {
    if (isValidNonNegativeNumericString(capacityString)) {
      this.setErrorState("capacity", false);
      return true;
    }

    this.setErrorState("cellSize", true);
    return false;
  };

  validateCellSize = (cellSizeString: unknown): boolean => {
    if (isValidCellSizeString(cellSizeString)) {
      this.setErrorState("cellSize", false);
      return true;
    }

    this.setErrorState("cellSize", true);
    return false;
  };

  validateSpeed = (speedString: unknown): boolean => {
    if (isValidPositiveNumericString(speedString)) {
      this.setErrorState("speed", false);

      return true;
    }

    this.setErrorState("speed", true);
    return false;
  };

  setErrorState = (property: keyof IInputErrors, show: boolean): void => {
    this.setState((prevState) => {
      return {
        ...prevState,
        errors: {
          ...prevState.errors,
          [property]: {
            ...prevState.errors[property],
            show: show,
          },
        },
      };
    });
  };

  showError = (error: { show: boolean; msg: string }): React.ReactNode => {
    return error.show ? (
      <div className="error-container">
        <p className="error-msg">{error.msg}</p>
      </div>
    ) : null;
  };

  debounce<T>(
    callback: (value: T) => void,
    property: string,
    validator: (value: T) => boolean,
    delay = Form.inputDelay
  ): (value: T) => void {
    return (value) => {
      this.clearTimer();

      this.timer = setTimeout(() => {
        if (validator(value)) {
          callback(value);
          this.setState((prevState) => {
            return {
              ...prevState,
              [property]: Number(value),
            };
          });
        }

        this.timer = null;
      }, delay);
    };
  }

  onChangeFn = (
    onChange: (value: string) => void,
    property: string,
    validator: (value: string) => boolean
  ): Callback<React.ChangeEvent<HTMLInputElement>, void> => {
    const debounced = this.debounce(onChange, property, validator);

    return (e: React.ChangeEvent<HTMLInputElement>) => {
      this.setState((prevState) => {
        return {
          ...prevState,
          [property]: e.target.value,
        };
      });
      debounced(e.target.value);
    };
  };

  onButtonClickFn = <T,>(
    fn: () => T
  ): Callback<React.FormEvent<HTMLButtonElement>, T> => {
    return (e: React.FormEvent<HTMLButtonElement>) => {
      e.preventDefault();
      return fn();
    };
  };

  onBlurFn = <T,>(
    onChange: (value: string) => T,
    validator: (value: string) => T
  ): Callback<React.FocusEvent<HTMLInputElement>, T | void> => {
    return (e: React.FocusEvent<HTMLInputElement>) => {
      this.clearTimer();

      if (validator(e.target.value)) {
        return onChange(e.target.value);
      }
    };
  };

  clearTimer = () => {
    if (this.timer !== null) {
      clearTimeout(this.timer);
      this.timer = null;
    }
  };

  componentWillUnmount() {
    this.clearTimer();
  }

  render() {
    return (
      <FormElement ref={this.formRef} data-testid={"field-form"}>
        <input autoComplete="off" hidden />
        <FormGroup>
          <FormField>
            <FormLabel htmlFor="field-width">Макс. ширина:</FormLabel>
            <FormInput
              id="field-width"
              type="number"
              pattern={floatPattern}
              step="1"
              name="field-width"
              autoFocus={true}
              value={this.state.maxFieldWidthString}
              autoComplete="off"
              onChange={this.onChangeFn(
                this.props.onMaxFieldWidthChange,
                "maxFieldWidthString",
                this.validateFieldWidth
              )}
              onBlur={this.onBlurFn(
                this.props.onMaxFieldWidthChange,
                this.validateFieldWidth
              )}
            />
            {this.showError(this.state.errors.fieldWidth)}
          </FormField>

          <FormField>
            <FormLabel htmlFor="field-height">Макс. высота:</FormLabel>
            <FormInput
              id="field-height"
              type="number"
              pattern={floatPattern}
              step="1"
              name="field-height"
              value={this.state.maxFieldHeightString}
              autoComplete="off"
              onChange={this.onChangeFn(
                this.props.onMaxFieldHeightChange,
                "maxFieldHeightString",
                this.validateFieldHeight
              )}
              onBlur={this.onBlurFn(
                this.props.onMaxFieldHeightChange,
                this.validateFieldHeight
              )}
            />
            {this.showError(this.state.errors.fieldHeight)}
          </FormField>

          <FormField>
            <FormLabel htmlFor="cell-size-input">Размер ячейки:</FormLabel>
            <FormInput
              id="cell-size-input"
              type="number"
              pattern={numberPattern}
              step="1"
              name="cell-size-input"
              value={this.state.cellSizeString}
              autoComplete="off"
              onChange={this.onChangeFn(
                this.props.onCellSizeChange,
                "cellSizeString",
                this.validateCellSize
              )}
              onBlur={this.onBlurFn(
                this.props.onCellSizeChange,
                this.validateCellSize
              )}
            />
            {this.showError(this.state.errors.cellSize)}
          </FormField>

          <FormField>
            <FormLabel htmlFor="capacity-percentage">
              Процент заполненности:
            </FormLabel>
            <p>{this.state.capacityString} %</p>
            <FormInput
              id="capacity-percentage"
              type="range"
              max="100"
              min="0"
              step="1"
              name="capacity-percentage"
              value={this.state.capacityString}
              autoComplete="off"
              onChange={this.onChangeFn(
                this.props.onCapacityChange,
                "capacityString",
                this.validateCapacity
              )}
              onBlur={this.onBlurFn(
                this.props.onCapacityChange,
                this.validateCapacity
              )}
            />
            {this.showError(this.state.errors.capacity)}
          </FormField>

          <FormField>
            <FormLabel htmlFor="speed-change">Обновлений в секунду:</FormLabel>
            <FormInput
              id="speed-change"
              type="number"
              pattern={floatPattern}
              step="0.1"
              name="speed-change"
              value={this.state.speedString}
              autoComplete="off"
              onChange={this.onChangeFn(
                this.props.onSpeedChange,
                "speedString",
                this.validateSpeed
              )}
              onBlur={this.onBlurFn(
                this.props.onSpeedChange,
                this.validateSpeed
              )}
            />
            {this.showError(this.state.errors.speed)}
          </FormField>
        </FormGroup>

        <FormGroup>
          <FormField>
            <FormButtonContainer>
              <FormButton
                type="button"
                onClick={this.onButtonClickFn(this.props.onStart)}
                name="start-button"
              >
                Старт
              </FormButton>
            </FormButtonContainer>
          </FormField>

          <FormField>
            <FormButtonContainer>
              <FormButton
                type="button"
                onClick={this.onButtonClickFn(this.props.onStop)}
                name="stop-button"
              >
                Стоп
              </FormButton>
            </FormButtonContainer>
          </FormField>

          <FormField>
            <FormButtonContainer>
              <FormButton
                type="reset"
                onClick={this.onButtonClickFn(this.props.onReset)}
                name="reset-button"
              >
                Сброс
              </FormButton>
            </FormButtonContainer>
          </FormField>
        </FormGroup>
      </FormElement>
    );
  }
}

const FormElement = styled.form`
  margin: 20px;
  display: flex;
  flex-direction: column;
  border: 1px solid black;
  border-radius: 10px;

  input,
  label,
  button,
  p {
    line-height: 1.2;
    font-size: 1.2em;
  }
`;

const FormGroup = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  max-width: 600px;
`;

const FormField = styled.div`
  display: inline-flex;
  flex-direction: column;
  flex: 0 0 calc(100% / 3 - 40px);
  margin: 20px;
  justify-content: end;

  input,
  button {
    margin: 5px 0 0 0;
    min-width: 100px;
  }

  div.error-container {
    position: relative;
  }

  p.error-msg {
    position: absolute;
    line-height: 1em;
    font-size: 1em;
    color: red;
  }
`;

const FormLabel = styled.label`
  display: inline;
`;

const FormInput = styled.input`
  background: inherit;
  border-top: 0;
  border-left: 0;
  border-right: 0;
  display: block;
  outline: 0;
  width: 100px;

  &:focus {
    border-bottom: 2px solid ${COLORS.accent};
  }
`;

const FormButtonContainer = styled.div`
  display: inline-block;
`;

const FormButton = styled.button`
  background: inherit;

  &:focus {
    outline: 2px solid ${COLORS.accent};
    border: 2px solid transparent;
  }
`;
