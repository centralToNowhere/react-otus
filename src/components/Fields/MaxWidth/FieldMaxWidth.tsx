import React, { useEffect } from "react";
import { l10n } from "@/l10n/ru";
import { InputField, LabelField } from "@/components/Fields";
import { FormField } from "@/components/Form/FormField";
import { useOnChangeHandler, onBlurHandler } from "@/components/Fields";
import { FieldError } from "@/components/Fields/FieldError/FieldError";
import { FormContainer, IFieldProps } from "@/components/Form";
import { useDebounce } from "@/utils";
import {
  onRawBlurHandler,
  onRawChangeHandler,
  preventNegativeNumbers,
} from "@/components/Fields/FieldHandlers";

export const FieldMaxWidth: React.FC<
  IFieldProps<{
    rawMaxWidth: string;
    rawMaxHeight: string;
    rawCellSize: string;
    highlight: boolean;
  }>
> = (props) => {
  const onChangeDebounced = useDebounce<string>(
    useOnChangeHandler(props.onChange, props.formValidators, props.onRawChange),
    FormContainer.inputDelay
  );

  const onChange = onRawChangeHandler((value) => {
    props.onRawChange(value);
    onChangeDebounced(value);
  });

  const onBlur = onRawBlurHandler((value) => {
    onChangeDebounced.clear();
    onBlurHandler(props.onChange, props.formValidators)(value);
  });

  useEffect(
    () => {
      return () => {
        onChangeDebounced.clear();
      };
    },
    // Stryker disable next-line ArrayDeclaration
    [onChangeDebounced]
  );

  return (
    <FormField>
      <LabelField htmlFor="field-width">{l10n.maxWidthLabel}</LabelField>
      <InputField
        id="field-width"
        type="number"
        step="1"
        name="fieldWidth"
        autoFocus={true}
        value={props.formRawData.rawMaxWidth}
        autoComplete="off"
        onChange={onChange}
        onKeyDown={preventNegativeNumbers}
        onBlur={onBlur}
        highlight={props.formRawData.highlight}
      />
      <FieldError show={props.error.show} msg={props.error.msg} />
    </FormField>
  );
};
