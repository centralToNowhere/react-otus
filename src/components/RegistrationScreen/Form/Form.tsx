import React, {FC, useCallback, MouseEvent, useState, SyntheticEvent} from "react";
import { l10n } from "@/l10n/ru";
import { AppAction } from "@/state";
import { IPlayer } from "@/state/actions";
import { useNavigate } from "react-router-dom";
import { FormField } from "@/components/Form/FormField";
import { FormElement } from "@/components/Form";
import { css } from "@emotion/react";
import { InputField, LabelField } from "@/components/Fields";
import { FormButton } from "@/components/Buttons";

interface IRegistrationFormProps {
  player: IPlayer,
  onPlayerRegistration: (playerName: string | null) => void
  dispatch: React.Dispatch<AppAction>
}

export const Form: FC<IRegistrationFormProps> = (props) => {
  const [playerName, setPlayerName] = useState(props.player.name);
  const navigate = useNavigate();

  const onPlayerNameChange = useCallback((e: SyntheticEvent<HTMLInputElement>) => {
    setPlayerName(e.currentTarget.value);
  }, [setPlayerName]);

  const registerPlayer = (e: MouseEvent<HTMLButtonElement>) => {
    props.onPlayerRegistration(playerName);
    navigate("/", { replace: true});
  };

  return (
    <FormElement css={css`
      background: #e8d5f0;
      width: 70%;
      max-width: 600px;
      display: inline-block;
    `}>
      <FormField
        css={css`
          display: flex;
        `}
      >
        <div css={css`
          margin-bottom: 50px;
        `}>
          <LabelField
            htmlFor="register-player-input" >{l10n.registerPlayerLabel}
          </LabelField>
          <InputField
            css={css`
            width: 100%;
          `}
            id={"register-player-input"}
            value={playerName || ""}
            onChange={onPlayerNameChange}
            aria-label="playerName"
            name="playerName"
            autoFocus={true}
          />
        </div>
        <FormButton
          type="button"
          css={css`
            margin-right: auto !important;
          `}
          name="startGameAsPlayer"
          onClick={registerPlayer}
        >
          {l10n.buttonStartGameAsPlayer}
        </FormButton>
      </FormField>
    </FormElement>
  )
}