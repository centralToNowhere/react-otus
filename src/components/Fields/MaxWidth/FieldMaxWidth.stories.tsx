import React from "react";
import { Story, Meta } from "@storybook/react";
import { FieldMaxWidth } from "./FieldMaxWidth";
import { IFieldProps } from "@/components/Form";
import { Provider } from "react-redux";
import { store } from "@/store/redux/store";
import { isValidPositiveNumericString } from "@/utils";

export default {
  title: "Components/FormFields/MaxWidth",
  component: FieldMaxWidth,
  args: {
    formRawData: {
      rawMaxWidth: "500",
      rawMaxHeight: "500",
      rawCellSize: "40",
      highlight: false,
    },
    onRawChange: () => {
      // empty function
    },
    onChange: () => {
      // empty function
    },
    formValidators: [isValidPositiveNumericString()],
    error: {
      show: false,
      msg: "maxWidth error",
    },
  },
} as Meta;

const Template: Story<
  IFieldProps<{
    rawMaxWidth: string;
    rawMaxHeight: string;
    rawCellSize: string;
    highlight: boolean;
  }>
> = (args) => {
  return (
    <Provider store={store}>
      <FieldMaxWidth {...args} />
    </Provider>
  );
};

export const MaxWidthInput = Template.bind({});
