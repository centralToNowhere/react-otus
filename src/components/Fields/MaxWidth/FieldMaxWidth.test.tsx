import React from "react";
import { render, screen, waitFor } from "@/utils/test-utils";
import { FieldMaxWidth } from "@/components/Fields/MaxWidth/FieldMaxWidth";
import { l10n } from "@/l10n/ru";
import { FormContainer } from "@/components/Form";
import userEvent from "@testing-library/user-event/dist";

jest.mock("@/Cell/Cell");

const cellMocks = jest.requireMock("@/Cell/Cell");
const cellOrigin = jest.requireActual("@/Cell/Cell");

const originalInputDelay = FormContainer.inputDelay;

beforeEach(() => {
  FormContainer.inputDelay = 50;
});

afterEach(() => {
  FormContainer.inputDelay = originalInputDelay;
  cellMocks.maxCellsAmount = cellOrigin.maxCellsAmount;
});

describe("FieldMaxWidth tests", () => {
  it("should render max width input", () => {
    const { asFragment } = render(
      <FieldMaxWidth
        value={"500"}
        onChange={() => {
          // empty function
        }}
      />,
      {
        preloadedState: {},
      }
    );

    const input: HTMLInputElement = screen.getByLabelText(l10n.maxWidthLabel);

    expect(input).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  describe("should call onChange", () => {
    [
      {
        maxFieldWidth: "1",
        maxFieldHeight: 100,
        cellSize: 1,
        maxCellsAmount: 100,
      },
      {
        maxFieldWidth: "10",
        maxFieldHeight: 100,
        cellSize: 10,
        maxCellsAmount: 10,
      },
      {
        maxFieldWidth: "100",
        maxFieldHeight: 100,
        cellSize: 10,
        maxCellsAmount: 100,
      },
      {
        maxFieldWidth: "1000",
        maxFieldHeight: 100,
        cellSize: 100,
        maxCellsAmount: 10,
      },
    ].forEach((settings) => {
      it(`width: ${settings.maxFieldWidth},
      height: ${settings.maxFieldHeight},
      cellSize: ${settings.cellSize},
      maxCellsAmount: ${settings.maxCellsAmount}`, async () => {
        cellMocks.maxCellsAmount = settings.maxCellsAmount;
        const onChange = jest.fn();

        render(<FieldMaxWidth value={"500"} onChange={onChange} />, {
          preloadedState: {
            fieldControl: {
              maxFieldWidth: 500,
              maxFieldHeight: settings.maxFieldHeight,
              cellSize: settings.cellSize,
              capacity: 100,
              speed: 2,
            },
          },
        });

        const input: HTMLInputElement = screen.getByLabelText(
          l10n.maxWidthLabel
        );

        userEvent.clear(input);
        userEvent.type(input, settings.maxFieldWidth);

        await waitFor(
          () => {
            expect(onChange).toHaveBeenCalledTimes(1);
          },
          {
            timeout: FormContainer.inputDelay + 100,
          }
        );
      });
    });
  });

  describe("should not call onChange", () => {
    [
      {
        maxFieldWidth: "-1",
        maxFieldHeight: 100,
        cellSize: 1,
        maxCellsAmount: 100,
      },
      {
        maxFieldWidth: "",
        maxFieldHeight: 100,
        cellSize: 1,
        maxCellsAmount: 100,
      },
      {
        maxFieldWidth: "qwerty",
        maxFieldHeight: 100,
        cellSize: 1,
        maxCellsAmount: 100,
      },
      {
        maxFieldWidth: "300",
        maxFieldHeight: 100,
        cellSize: 5,
        maxCellsAmount: 100,
      },
      {
        maxFieldWidth: "0",
        maxFieldHeight: 100,
        cellSize: 100,
        maxCellsAmount: 100,
      },
      {
        maxFieldWidth: "99",
        maxFieldHeight: 100,
        cellSize: 100,
        maxCellsAmount: 100,
      },
    ].forEach((settings) => {
      it(`width: ${settings.maxFieldWidth},
      height: ${settings.maxFieldHeight},
      cellSize: ${settings.cellSize},
      maxCellsAmount: ${settings.maxCellsAmount}`, async () => {
        cellMocks.maxCellsAmount = settings.maxCellsAmount;
        const onChange = jest.fn();

        render(<FieldMaxWidth value={"500"} onChange={onChange} />, {
          preloadedState: {
            fieldControl: {
              maxFieldWidth: 500,
              maxFieldHeight: settings.maxFieldHeight,
              cellSize: settings.cellSize,
              capacity: 100,
              speed: 2,
            },
          },
        });

        const input: HTMLInputElement = screen.getByLabelText(
          l10n.maxWidthLabel
        );

        userEvent.clear(input);
        userEvent.type(input, settings.maxFieldWidth);

        await new Promise<void>((resolve) => {
          setTimeout(() => {
            resolve();
          }, FormContainer.inputDelay + 100);
        });

        expect(onChange).not.toHaveBeenCalled();
      });
    });
  });
});
