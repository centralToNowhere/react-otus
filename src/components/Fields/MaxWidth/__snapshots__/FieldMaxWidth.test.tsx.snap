// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldMaxWidth tests should render max width input 1`] = `
<DocumentFragment>
  <div
    class="css-w6s8l8-FormField e1jhpckf0"
  >
    <label
      class="css-m0dhjx-LabelField e16iqet90"
      for="field-width"
    >
      Макс. ширина:
    </label>
    <input
      autocomplete="off"
      class="css-iol0j7-InputField el9r7ir0"
      id="field-width"
      name="fieldWidth"
      pattern="([0-9]*[.])?[0-9]+"
      step="1"
      type="number"
      value="500"
    />
  </div>
</DocumentFragment>
`;
